Async chain

bdeggleston opened the pull request:
This PR relates to https://github.com/apache/cassandra-accord/pull/17

maedhroz committed:
CEP-15: Multi-Partition Transaction CQL Support (Alpha)

patch by Caleb Rackliffe; reviewed by David Capwell for CASSANDRA-17719

Co-authored-by: Caleb Rackliffe <calebrackliffe@gmail.com>
Co-authored-by: Blake Eggleston <bdeggleston@gmail.com>
Co-authored-by: David Capwell <dcapwell@apache.org>

belliottsmith committed:
Partial Replication of Transaction State

The replication of transactions and their processing state must be optimal, i.e. linearly proportional to the size of the transaction and the number of keys involved. This patch ensures a transaction is replicated in slices, with each slice being written (replication factor) times only to the shard that owns the keys, and only once (except in the case the transaction is recovered).

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-17109

bdeggleston committed:
CEP-15 (C*): Messaging and storage engine integration

patch by Blake Eggleston; reviewed by Benedict Elliott Smith, David Capwell for CASSANDRA-17103

belliottsmith committed:
CEP-15: Routables
  - Integrate accord-core changes for CASSANDRA-18087

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18087

dcapwell committed:
Command table now uses a local version added to the payload rather than rely on a version column, and fixed a few feedback related issues

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-17103

belliottsmith committed:
Transaction Expiration

Ensure old or abandoned transactions are invalidated wherever possible.

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18041

belliottsmith committed:
Invalidation fixes/improvements
 - Integrate accord-core changes for CASSANDRA-18057

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18057

belliottsmith committed:
Refactor response tracking to improve efficiency and clarity; introduce dedicated property tests; re-activate fast-path during range movements

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18056

aweisberg committed:
Support CAS and serial read on Accord

patch by Ariel Weisberg; Reviewed by Caleb Rackliffe for CASSANDRA-18100

dcapwell committed:
Ninja for CASSANDRA-17719: Adder/Substraction should return NULL if either the current or the user value are NULL

dcapwell committed:
Ninja for CASSANDRA-17719: Add @Simulate(with = MONITORS) to MultiReadFuture to get simulator working

dcapwell committed:
Ninja for CASSANDRA-17719: When a reference sees a null, return Constants.NULL_VALUE rather than try to parse it

dcapwell committed:
Ninja for CASSANDRA-17719: TransactionStatement now logs every exception, will revert before trunk but needed for debugging

dcapwell committed:
Ninja for CASSANDRA-17719: When AccordCommand.setPartialTxn is called,
make sure to update this.kind in order to make tests stable

dcapwell committed:
Ninja for CASSANDRA-17719: Changed
AsyncWriterTest#commandsPerKeyDenormalization to use SaveStatus rather
than Status

dcapwell committed:
Ninja for CASSANDRA-17719 disable the checks in increaseSlightly and decreaseSlightly to get back to the old behavior for now

dcapwell committed:
Ninja for CASSANDRA-17719: accord.primitives.Range#someIntersectingRoutingKey was added but does not work in all cases in C* due to sentinal values, added logic to return a C* friendly token

dcapwell committed:
Refactor AccordTestBase to block retries on non-idempotent transactions.
Some tests may be flaky now due to Preempted being thrown.

dcapwell commented:
FYI need the following so tests can run

```
diff --git .build/include-accord.sh .build/include-accord.sh
index 0dc01bf2ba..079c96da1c 100755
--- .build/include-accord.sh
+++ .build/include-accord.sh
@@ -24,8 +24,8 @@ set -o nounset

 bin="$(cd "$(dirname "$0")" > /dev/null; pwd)"

-accord_repo='https://github.com/apache/cassandra-accord.git'
-accord_branch='trunk'
+accord_repo='https://github.com/bdeggleston/cassandra-accord.git'
+accord_branch='async-chain'
 accord_src="$bin/cassandra-accord"

 checkout() {
```

dcapwell commented on src/java/org/apache/cassandra/service/accord/AccordCommand.java:
```diff
@@ -652,59 +654,66 @@ private boolean canApplyWithCurrentScope(SafeCommandStore safeStore)
     }
 
-    private Future<Void> applyWithCorrectScope(CommandStore unsafeStore)
+    private AsyncResult<Void> applyWithCorrectScope(CommandStore unsafeStore)
```
The usage is `AsyncChain`, we can simplify to be

```
private AsyncChain<Void> applyWithCorrectScope(CommandStore unsafeStore)
    {
        TxnId txnId = txnId();
        return unsafeStore.submit(this, safeStore -> {
            AccordCommand command = (AccordCommand) safeStore.command(txnId);
            return command.applyChain(safeStore, false);
        }).flatMap(a -> a);
    }
```

I don't like `.flatMap(a -> a);` but feel this is sadly a java thing... we can have `.flatMap(AsyncChains.flatten())` but that's more verbose...  

If this returns a `AsyncResult` because we are trying to control execution order, then can just add `.beginAsResult()` to the end of the `flatMap`

belliottsmith replied:
Agreed, I applied the same modification in my branch. It’s not optimal, but it’s probably better than other approaches (another option is to offer a flatMap function in CommandStore, but it would likely be implemented the same way)

dcapwell commented on src/java/org/apache/cassandra/service/accord/async/AsyncOperation.java:
```diff
@@ -32,11 +33,12 @@
+import accord.utils.async.AsyncChains;
 import org.apache.cassandra.service.accord.AccordCommandStore;
 import org.apache.cassandra.service.accord.api.PartitionKey;
 import org.apache.cassandra.utils.concurrent.AsyncPromise;
```
unused import

dcapwell commented on src/java/org/apache/cassandra/service/accord/async/AsyncOperation.java:
```diff
@@ -208,6 +221,14 @@ public void run()
     }
 
+    @Override
+    public void begin(BiConsumer<? super R, Throwable> callback)
```
`AsyncChains.Head` doesn't have exactly-once semantics, which can put us into a situation where we double invoke.  Before this patch `org.apache.cassandra.service.accord.AccordCommandStore#submit` controlled this behavior by always returning the result (Future), but now any caller to `submit` will get a lazy `AsyncChain` and if those callers touch `begin` twice this could become an issue.

I think most usages of `AsyncChains.Head` would benefit from adding exactly-once semantics.

dcapwell replied:
Here is a quick test to show this example

```
@Test
    void beginMultipleTimes()
    {
        final AtomicBoolean called = new AtomicBoolean(false);
        Supplier<Integer> task = () -> {
            if (!called.compareAndSet(false, true))
                throw new AssertionError("Attempted to invoke multiple times");
            return 5;
        };
        // randomly generate different possible cases where we could take a callable
        Gen<AsyncChain<Integer>> gen = rnd -> {
            called.set(false);
            if (rnd.nextBoolean())
                return AsyncChains.ofCallable(MoreExecutors.directExecutor(), () -> task.get());
            return new AsyncChains.Head<>() {
                @Override
                public void begin(BiConsumer<? super Integer, Throwable> callback) {
                    callback.accept(task.get(), null);
                }
            };
        };
        qt().forAll(gen).check(chain -> {
            Assertions.assertEquals(5, Result.from(chain).get());

            chain = chain.map(i -> i + 1);

            AsyncChain<String> a = chain.map(Object::toString);

            Assertions.assertEquals("6", Result.from(a).get());
            Assertions.assertEquals(7, Result.from(a.map(Integer::parseInt).map(i -> i + 1)).get());
        });
    }
```

this fails with the stack trace

```
Caused by: java.lang.AssertionError: Attempted to invoke multiple times
	at accord.utils.async.AsyncChainsTest.lambda$test$4(AsyncChainsTest.java:101)
	at accord.utils.async.AsyncChainsTest.lambda$test$5(AsyncChainsTest.java:107)
	at accord.utils.async.AsyncChains.lambda$encapsulate$0(AsyncChains.java:272)
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30)
	at accord.utils.async.AsyncChains$1.begin(AsyncChains.java:314)
	at accord.utils.async.AsyncChains$Head.begin(AsyncChains.java:96)
	at accord.utils.async.AsyncChains$Link.begin(AsyncChains.java:121)
	at accord.utils.async.AsyncChains$Map.begin(AsyncChains.java:125)
	at accord.utils.async.AsyncChainsTest$Result.from(AsyncChainsTest.java:151)
```

Now, if you comment out the first statement `Assertions.assertEquals(5, Result.from(chain).get());` then the test fails as you attempt to `map` after you extracted the result, this is handled here

```
Caused by: java.lang.IllegalStateException
	at com.google.common.base.Preconditions.checkState(Preconditions.java:491)
	at accord.utils.async.AsyncChains.add(AsyncChains.java:260)
	at accord.utils.async.AsyncChains.map(AsyncChains.java:232)
	at accord.utils.async.AsyncChainsTest.lambda$beginMultipleTimes$7(AsyncChainsTest.java:107)
```

which is this check: `Preconditions.checkState(next instanceof Head<?>);`

belliottsmith replied:
I have fixed this in my branch

belliottsmith committed:
Shard local CommandStores on contiguous ranges

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18142

belliottsmith committed:
Refactor Timestamp/TxnId
 - Combine real and logical into a single 64-but HLC
 - Introduce 16 flag bits
 - Pack epoch (48-bits), HLC (64-bits) and flags (16-bits) into two longs in memory

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18172

belliottsmith committed:
Introduce RangeDeps

Refactor Deps into KeyDeps and RangeDeps

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18173

belliottsmith committed:
Introduce Range transactions

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18174

dcapwell committed:
Ninja: Add AccordTestUtils.parse which was missing in the latest commit

belliottsmith committed:
ninja: pick a cassandra-accord SHA to build against to decouple source trees

dcapwell committed:
CASSANDRA-18154: CEP-15: Enhance returning SELECT to allow partition and clustering IN clauses to return multiple partitions/rows

dcapwell committed:
CEP-15: (C*) Enhance returning SELECT to allow partition and clustering IN clauses to return multiple partitions/rows

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18154

belliottsmith committed:
ninja fix CASSANDRA-18142: System.distributedKeyspaces is unsorted

aweisberg committed:
Ninja fix DatabaseDescriptorRefTest

maedhroz committed:
Post-rebase fixes around CASSANDRA-18099 and CASSANDRA-18114 (among other things)

bdeggleston committed:
async-chain integration
